sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java

```diff
@@ -349,32 +411,34 @@ private CrateCollector getShardsCollector(RoutedCollectPhase collectPhase,
                     continue;
                 }
                 try {
-                    ShardCollectService shardCollectService =
-                        indexService.shardInjectorSafe(shard).getInstance(ShardCollectService.class);
-
-                    Object[] row = shardCollectService.getRowForShard(normalizedPhase);
+                    ShardId shardId = new ShardId(indexName, shard);
+                    ShardCollectorProvider shardCollectorProvider = getCollectorProviderSafe(shardId);
+                    Object[] row = shardCollectorProvider.getRowForShard(normalizedPhase);
                     if (row != null) {
                         rows.add(row);
                     }
                 } catch (ShardNotFoundException | IllegalIndexShardStateException e) {
                     unassignedShards.add(toUnassignedShard(new ShardId(indexName, shard)));
                 } catch (Throwable t) {
+                    t.printStackTrace();
                     throw new UnhandledServerException(t);
                 }
             }
         }
         if (!unassignedShards.isEmpty()) {
             // since unassigned shards aren't really on any node we use the collectPhase which is NOT normalized here
             // because otherwise if _node was also selected it would contain something which is wrong
-            for (Object[] objects : Iterables.transform(
-                systemCollectSource.toRowsIterable(collectPhase, unassignedShards, false), Row.MATERIALIZE)) {
+            for (Object[] objects : StreamSupport.stream(
+                systemCollectSource.toRowsIterable(collectPhase, unassignedShards, false).spliterator(), false)
+                .map(Row.MATERIALIZE::apply).collect(Collectors.toList())) {
```

Comment:
I think `.collect(Collectors.ToList)` isn't really necessary here.  `rows.add(objects)` could be done with a `forEach` on the stream directly